src/scheduler/MpiWorld.cpp

```diff
@@ -714,17 +787,65 @@ void MpiWorld::awaitAsyncRequest(int requestId)
 {
     SPDLOG_TRACE("MPI - await {}", requestId);
 
-    auto it = futureMap.find(requestId);
-    if (it == futureMap.end()) {
-        throw std::runtime_error(
-          fmt::format("Error: waiting for unrecognized request {}", requestId));
+    // Get the corresponding send and recv ranks
+    auto it = reqIdToRanks.find(requestId);
+    if (it == reqIdToRanks.end()) {
+        SPDLOG_ERROR("Asynchronous request id not recognized: {}", requestId);
+        throw std::runtime_error("Unrecognized async request id");
+    }
+    int sendRank = it->second.first;
+    int recvRank = it->second.second;
+    reqIdToRanks.erase(it);
+
+    // If awaiting an isend request, return immediately as our transport layer
+    // is asynchronous, thus request is already handled
+    if (sendRank == MPI_IS_ISEND_REQUEST && recvRank == MPI_IS_ISEND_REQUEST) {
+        return;
     }
 
-    // This call blocks until requestId has finished.
-    it->second.wait();
-    futureMap.erase(it);
+    std::shared_ptr<faabric::scheduler::UnackedMessageBuffer> umb =
+      getUnackedMessageBuffer(sendRank, recvRank);
+
+    // The request id must be in the UMB, as an irecv must happen before an
+    // await. It is very likely that if the assert were to fail, the previous
+    // check would have also failed, thus why we assert here.
+    auto idIndex = std::find(umb->ids.begin(), umb->ids.end(), requestId);
+    assert(idIndex != umb->ids.end());
+    auto idDistance = std::distance(umb->ids.begin(), idIndex);
+
+    // Get the corresponding message
+    assert(idDistance < umb->msgs.size());
+    auto msgIndex = umb->msgs.begin();
+    std::advance(msgIndex, idDistance);
+
+    // Get the corresponding request arguments
+    assert(idDistance < umb->args.size());
+    auto argsIndex = umb->args.begin();
+    std::advance(argsIndex, idDistance);
```

Comment:
This logic using `advance` and `distance` is quite hard to read but seems like it's all operations on the message buffer, therefore could this not be a method on that object? E.g. `umb.ackMessage(requestId)`?